ledger/heavy/replica/target.go

```diff
@@ -0,0 +1,184 @@
+
+// Target provides methods to get notification by subscription.
+type Target interface {
+	// Notify informs target replica node.
+	Notify(context.Context, insolar.PulseNumber) error
+}
+
+type Page struct {
+	Scope byte
+	Skip  uint32
+	Limit uint32
+	Pulse insolar.PulseNumber
+}
+
+func NewTarget(cfg configuration.Replica, parent Parent) Target {
+	scopes := append(cfg.ScopesToReplicate, byte(store.ScopePulse))
+	return &target{cfg: cfg, parent: parent, scopesToReplicate: scopes}
+}
+
+type target struct {
+	Sequencer         sequence.Sequencer  `inject:""`
+	JetKeeper         JetKeeper           `inject:""`
+	Validator         integrity.Validator `inject:""`
+	cfg               configuration.Replica
+	parent            Parent
+	scopesToReplicate []byte
+}
+
+func (t *target) Start(ctx context.Context) error {
+	if t == nil {
+		return errors.New("invalid target component")
+	}
+	pn := t.JetKeeper.TopSyncPulse()
+	at := Page{Pulse: pn}
+	go t.subscribe(at)
+	return nil
+}
+
+func (t *target) Notify(ctx context.Context, present insolar.PulseNumber) error {
+	go t.process(present)
+	return nil
+}
+
+func (t *target) subscribe(at Page) {
+	ctx := context.Background()
+	logger := inslogger.FromContext(ctx)
+	logger.Debugf("target.subscribe at pulse=%v", at.Pulse)
+	for i := 0; i < t.cfg.Attempts; i++ {
+		err := t.parent.Subscribe(ctx, t, at)
+		if err != nil {
+			logger.Error(err)
+			time.Sleep(t.cfg.DelayForAttempt)
+			continue
+		}
+		return
+	}
+	logger.Errorf("Failed to subscribe to parent replica. The maximum number of attempts is exceeded.")
+}
+
+func (t *target) process(present insolar.PulseNumber) {
+	next := genesis()
+	synced := t.JetKeeper.TopSyncPulse()
+	if synced != genesis() {
+		next = t.nextPulse(synced)
+	}
+
+	for next <= present {
+		if !t.fetch(next) || !t.finish(next) {
+			return
+		}
+		curr := next
+		next = t.nextPulse(next)
+		if next == curr {
+			break
+		}
+	}
+
+	go t.subscribe(Page{Pulse: next})
+}
+
+func genesis() insolar.PulseNumber {
+	return insolar.GenesisPulse.PulseNumber
+}
+
+func (t *target) nextPulse(pn insolar.PulseNumber) insolar.PulseNumber {
+	pulses := t.Sequencer.Slice(byte(store.ScopePulse), pn, 0, 1)
+	if len(pulses) == 0 {
+		return pn
+	}
+	pulseItem := pulseNode{}
+	if err := insolar.Deserialize(pulses[0].Value, &pulseItem); err != nil {
+		return pn
+	}
+	return *pulseItem.Next
+}
+
+func (t *target) fetch(need insolar.PulseNumber) bool {
+	for _, scope := range t.scopesToReplicate {
+		if !t.pull(scope, need) {
+			t.subscribe(Page{Pulse: need})
+			return false
+		}
+	}
+	return true
+}
+
+func (t *target) pull(scope byte, pn insolar.PulseNumber) bool {
+	ctx := context.Background()
+	logger := inslogger.FromContext(ctx)
+	skip := t.Sequencer.Len(scope, pn)
+	for {
+		at := Page{Scope: scope, Skip: skip, Limit: t.cfg.DefaultBatchSize, Pulse: pn}
+		packet, total, err := t.parent.Pull(ctx, at)
```

Comment:
Who is the paren? if this should run on a heavy node? We have a db, we can just read from it.

Reply:
Parent is a just abstraction of the source of replication data. And at most cases, it will be a remote heavy node.